In [24]:
import pandas as pd
import re
import nltk
nltk.download('stopwords')
nltk.download('wordnet')
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\PRATAP\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\PRATAP\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\wordnet.zip.


In [10]:
message = pd.read_csv('./Dataset/SMSSpamCollection', sep = '\t', names = ["label", "message"])

In [11]:
message.head(10)

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...
5  spam  FreeMsg Hey there darling it's been 3 week's n...
6   ham  Even my brother is not like to speak with me. ...
7   ham  As per your request 'Melle Melle (Oru Minnamin...
8  spam  WINNER!! As a valued network customer you have...
9  spam  Had your mobile 11 months or more? U R entitle...

In [14]:
wordNetLemmatizer = WordNetLemmatizer()
corpus = []

In [19]:
message['message'][0]

'Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'

In [25]:
for i in range(0, len(message)) :
    review = re.sub('[^A-Za-z]', ' ', message['message'][i])
    review = review.lower()
    review = review.split()
    review = [wordNetLemmatizer.lemmatize(word) for word in review if word not in stopwords.words('english')]
    review = ' '.join(review)
    corpus.append(review)

In [43]:
from sklearn.feature_extraction.text import CountVectorizer

cv =  CountVectorizer(max_features = 5000)
X = cv.fit_transform(corpus).toarray()

In [ ]:
y = pd.get_dummies(message['label'])
y = y.iloc[:, 1].values

In [56]:
print(X.shape)
print(y.shape)

(5572, 5000)
(5572,)


In [46]:
len(y)

5572

In [47]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [48]:
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB()
model.fit(X_train, y_train)

MultinomialNB()

In [49]:
y_pred = model.predict(X_test)

In [50]:
y_pred

array([0, 1, 0, ..., 0, 1, 0], dtype=uint8)

In [51]:
y_test

array([0, 1, 0, ..., 0, 1, 0], dtype=uint8)

In [52]:
from sklearn.metrics import confusion_matrix

matrix = confusion_matrix(y_test, y_pred)

In [53]:
matrix

array([[942,  13],
       [  8, 152]], dtype=int64)

# So, from the prediction, we can see that out of 1115 rows, 1094 rows(952 + 152) have been correctly predicted

In [57]:
from sklearn.metrics import accuracy_score

accuracy_score = accuracy_score(y_test, y_pred)
print(accuracy_score)

0.9811659192825112


# So the accuracy is around 98%